vllm/spec_decode/multi_step_worker.py

```diff
@@ -185,26 +186,33 @@ def _filter_model_output(
             List[SamplerOutput]: A list containing the filtered model 
             outputs for the specified indices.
         """
-        return [
-            SamplerOutput(
-                outputs=[
-                    expanded_batch_output.outputs[i]
-                    for i in output_indices_to_retain
-                ] if len(expanded_batch_output.outputs) > 0 else [],
-                sampled_token_probs=(
-                    expanded_batch_output.
-                    sampled_token_probs[output_indices_to_retain]
-                    if expanded_batch_output.sampled_token_probs is not None
-                    else None),
-                logprobs=(
-                    expanded_batch_output.logprobs[output_indices_to_retain]
-                    if expanded_batch_output.logprobs is not None else None),
-                sampled_token_ids=(expanded_batch_output.
-                                   sampled_token_ids[output_indices_to_retain]
-                                   if expanded_batch_output.sampled_token_ids
-                                   is not None else None))
-            for expanded_batch_output in expanded_batch_outputs
-        ]
+        def _filter_optional_tensor(tensor: torch.Tensor, indices: torch.Tensor) -> torch.Tensor:
+            return tensor.index_select(0, indices) if tensor is not None else None
+
+        if not expanded_batch_outputs:
+            return []
+
+        output_indices_to_retain_tensor = torch.tensor(output_indices_to_retain, device=device)
+
+        filtered_outputs = []
+        for expanded_batch_output in expanded_batch_outputs:
+            outputs = (
+                [expanded_batch_output.outputs[i] for i in output_indices_to_retain]
+                if expanded_batch_output.outputs else []
+            )
+
+            sampled_token_probs = _filter_optional_tensor(expanded_batch_output.sampled_token_probs, output_indices_to_retain_tensor)
+            logprobs = _filter_optional_tensor(expanded_batch_output.logprobs, output_indices_to_retain_tensor)
+            sampled_token_ids = _filter_optional_tensor(expanded_batch_output.sampled_token_ids, output_indices_to_retain_tensor)
+
+            filtered_outputs.append(SamplerOutput(
+                outputs=outputs,
+                sampled_token_probs=sampled_token_probs,
+                logprobs=logprobs,
+                sampled_token_ids=sampled_token_ids
+            ))
+
+        return filtered_outputs
```

Comment:
@jacob-crux I'm not sure the existing code needs to be rewritten, wouldn't it be sufficient to just add this at the start?
```python
        if not expanded_batch_outputs:
            return []

        output_indices_to_retain = torch.tensor(output_indices_to_retain, device=device)
```

Reply:
Please refer to the speed difference between list slicing and `tensor.index_select` that I have profiled. 
Since this part does not significantly impact performance, we can implement the change by simply modifying the `if statement` and converting `output_indices_to_retain` to a tensor, as you mentioned.